synapse/events/spamcheck.py

```diff
@@ -211,33 +301,28 @@ async def check_registration_for_spam(
             Enum for how the request should be handled
         """
 
-        for spam_checker in self.spam_checkers:
-            # For backwards compatibility, only run if the method exists on the
-            # spam checker
-            checker = getattr(spam_checker, "check_registration_for_spam", None)
-            if checker:
-                # Provide auth_provider_id if the function supports it
-                checker_args = inspect.signature(checker)
-                if len(checker_args.parameters) == 4:
-                    d = checker(
-                        email_threepid,
-                        username,
-                        request_info,
-                        auth_provider_id,
-                    )
-                elif len(checker_args.parameters) == 3:
-                    d = checker(email_threepid, username, request_info)
-                else:
-                    logger.error(
-                        "Invalid signature for %s.check_registration_for_spam. Denying registration",
-                        spam_checker.__module__,
-                    )
-                    return RegistrationBehaviour.DENY
-
-                behaviour = await maybe_awaitable(d)
-                assert isinstance(behaviour, RegistrationBehaviour)
-                if behaviour != RegistrationBehaviour.ALLOW:
-                    return behaviour
+        for checker in self._callbacks.get("check_registration_for_spam", []):
+            # Provide auth_provider_id if the function supports it
+            checker_args = inspect.signature(checker)
+            if len(checker_args.parameters) == 4:
+                d = checker(
+                    email_threepid,
+                    username,
+                    request_info,
+                    auth_provider_id,
+                )
+            elif len(checker_args.parameters) == 3:
+                d = checker(email_threepid, username, request_info)
+            else:
+                logger.error(
+                    "Invalid signature for check_registration_for_spam. Denying registration",
+                )
+                return RegistrationBehaviour.DENY
```

Comment:
you could sort this out in `register_callback`: if the callback only takes three params, wrap it with a function which takes 4.